File: src/0221_maximal_square.py

# LeetCode 221: Maximal Square
# Python version

def maximalSquare(matrix: list[list[str]]) -> int:
    if not matrix or not matrix[0]:
        return 0

    rows, cols = len(matrix), len(matrix[0])
    prev = [0] * (cols + 1)
    best = 0

    for r in range(rows - 1, -1, -1):
        curr = [0] * (cols + 1)
        for c in range(cols - 1, -1, -1):
            if matrix[r][c] == "1":
                curr[c] = 1 + min(prev[c], curr[c + 1], prev[c + 1])
                best = max(best, curr[c])
        prev = curr

    return best * best



def main():
    assert maximalSquare([['1', '0', '1', '0', '0'], ['1', '0', '1', '1', '1'], ['1', '1', '1', '1', '1'], ['1', '0', '0', '1', '0']]) == 4
    assert maximalSquare([['0', '1'], ['1', '0']]) == 1
    assert maximalSquare([['0']]) == 0

if __name__ == "__main__":
    main()

File: src/0221_maximal_square.rs
pub fn maximal_square(matrix: Vec<Vec<String>>) -> i32 {
    if matrix.is_empty() || matrix[0].is_empty() {
        return 0;
    }

    let rows = matrix.len();
    let cols = matrix[0].len();
    let mut prev = vec![0; cols + 1];
    let mut best = 0;

    for r in (0..rows).rev() {
        let mut curr = vec![0; cols + 1];
        for c in (0..cols).rev() {
            if matrix[r][c] == "1" {
                curr[c] = 1 + prev[c].min(curr[c + 1]).min(prev[c + 1]);
                best = best.max(curr[c]);
            }
        }
        prev = curr;
    }

    best * best
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            maximal_square(vec![
                vec![
                    "1".to_string(),
                    "0".to_string(),
                    "1".to_string(),
                    "0".to_string(),
                    "0".to_string()
                ],
                vec![
                    "1".to_string(),
                    "0".to_string(),
                    "1".to_string(),
                    "1".to_string(),
                    "1".to_string()
                ],
                vec![
                    "1".to_string(),
                    "1".to_string(),
                    "1".to_string(),
                    "1".to_string(),
                    "1".to_string()
                ],
                vec![
                    "1".to_string(),
                    "0".to_string(),
                    "0".to_string(),
                    "1".to_string(),
                    "0".to_string()
                ]
            ]),
            4
        );
        assert_eq!(
            maximal_square(vec![
                vec!["0".to_string(), "1".to_string()],
                vec!["1".to_string(), "0".to_string()]
            ]),
            1
        );
        assert_eq!(maximal_square(vec![vec!["0".to_string()]]), 0);
    }
}
